keras_core/utils/dataset_utils.py

```diff
@@ -6,44 +12,503 @@
         seed: A random seed for shuffling.
 
     Returns:
-        A tuple of two `tf.data.Dataset` objects: the left and right splits.
+        A tuple of two `tf.data.Dataset or torchDataset` objects:
+        the left and right splits.
 
     Example:
 
     >>> data = np.random.random(size=(1000, 4))
-    >>> left_ds, right_ds = split_dataset(data, left_size=0.8)
+    >>> left_ds, right_ds = keras_core.utils.split_dataset(data, left_size=0.8)
     >>> int(left_ds.cardinality())
     800
     >>> int(right_ds.cardinality())
     200
+
     """
-    # TODO: long-term, port implementation.
-    return tf.keras.utils.split_dataset(
-        dataset,
-        left_size=left_size,
-        right_size=right_size,
-        shuffle=shuffle,
-        seed=seed,
+    from torch.utils.data import Dataset as torchDataset
+
+    dataset_type_spec = _get_type_spec(dataset)
+
+    if dataset_type_spec not in [
+        torchDataset,
+        tf.data.Dataset,
+        list,
+        tuple,
+        np.ndarray,
+    ]:
+        raise TypeError(
+            "The `dataset` argument must be either a `tf.data.Dataset` "
+            "object or a list/tuple of arrays. "
+            f"Received: dataset={dataset} of type {type(dataset)}"
+        )
+
+    if right_size is None and left_size is None:
+        raise ValueError(
+            "At least one of the `left_size` or `right_size` "
+            "must be specified. Received: left_size=None and "
+            "right_size=None"
+        )
+
+    dataset_as_list = _convert_dataset_to_list(dataset, dataset_type_spec)
+
+    if shuffle:
+        if seed is None:
+            seed = random.randint(0, int(1e6))
+        random.seed(seed)
+        random.shuffle(dataset_as_list)
+
+    total_length = len(dataset_as_list)
+
+    left_size, right_size = _rescale_dataset_split_sizes(
+        left_size, right_size, total_length
+    )
+    left_split = list(dataset_as_list[:left_size])
+    right_split = list(dataset_as_list[-right_size:])
+
+    left_split = _restore_dataset_from_list(
+        left_split, dataset_type_spec, dataset
     )
+    right_split = _restore_dataset_from_list(
+        right_split, dataset_type_spec, dataset
+    )
+
+    if dataset_type_spec != torchDataset:
+        left_split = tf.data.Dataset.from_tensor_slices(left_split)
+        right_split = tf.data.Dataset.from_tensor_slices(right_split)
+
+        # apply batching to the splits if the dataset is batched
+        if dataset_type_spec is tf.data.Dataset and is_batched(dataset):
+            batch_size = get_batch_size(dataset)
+            if batch_size is not None:
+                left_split = left_split.batch(batch_size)
+                right_split = right_split.batch(batch_size)
+
+        left_split = left_split.prefetch(tf.data.AUTOTUNE)
+        right_split = right_split.prefetch(tf.data.AUTOTUNE)
+        return left_split, right_split
+
+    elif dataset_type_spec == torchDataset:
```

Comment:
On second thought, I don't think we should support torch datasets at all here, because the API is becoming completely inconsistent:

- pass numpy arrays, get back tf.data.Dataset
- pass tf.data.Dataset, get back tf.data.Dataset
- pass torch dataset, get back torch dataset

Let's just stick to always returning a tf.data.Dataset IMO.

Reply:
@fchollet but if that's the case how we this function can be used with torch workflow as it will require to have torchDataset ?

Reply:
@fchollet and since returning torch is only constrained within torch backend and will not impact other it should be safe as even going forward with jax we would need to add for that too. What do you think ? 🧐 Also if api consistency is issue one possible solution is to only return numpy arrays and will leave at user how it is needed, this way we will ensure to support every framework.

Reply:
`tf.data.Datasets` are supported by Keras models with all backends, so it's fine IMO. It's also the only way to stay be backwards compatible with tf.keras, which is very important.

Reply:
@fchollet aah got it then let's only use `tf.data.Datasets`

Reply:
@fchollet Please you can review now following changes are made.

- `torch.data.Dataset` is only imported once.
- return will be limited to tf.data.Datasets.
- fixed indentation.

@fchollet Need clarification from you  I have moved logic of detecting torch tensor and torch dataset to `is_torch_tensor` and `is_torch_dataset` since `is_torch_dataloader` is under trainers should I move these function there or keep it here only ?